### PR TITLE
Deprecate PYTROLL_CHUNK_SIZE and fallback to array.chunk-size config from dask

### PR DESCRIPTION
**Pull request by mraspaud**
This PR starts deprecating PYTROLL_CHUNK_SIZE in favour of dask's `array.chunk-size` setting.
This is a first step in migrating the readers to determine chunk size for the data based on the data itself instead of using square chunks by default. For example, data stored as stripes may load much faster if the horizontal striping is kept as much as possible instead of cutting the data in square chunks.
Individual adaptations need to be implemented for each reader in later PRs.

 - [x] Tests added

### CODE REVIEW

File: satpy/tests/test_utils.py
Comment (djhoese) [inline satpy/tests/test_utils.py:505]: I may be misunderstanding this, but why not `with dask.config.set(...):`?

Comment (mraspaud) [reply satpy/tests/test_utils.py:505]: ah, good point, I forgot that was a possibility

**Comment by mraspaud** [comment]
I need to work more on this, please don't merge yet.

**Comment by mraspaud** [comment]
> I need to work more on this, please don't merge yet.

Ready now

File: satpy/utils.py
Comment (djhoese) [inline satpy/utils.py:626]: `array.chunk-size` is in bytes/memory size right? So there should probably be some language in this warning about 32-bit float versus 64-bit versus integers, etc.

Comment (djhoese) [reply satpy/utils.py:626]: Also, do you want to specify this as a PendingDeprecationWarning or a regular DeprecationWarning? Or just leave it as a UserWarning?

File: satpy/utils.py
Comment (djhoese) [inline satpy/utils.py:615]: Can you swap the logic on this so the condition is `if chunk_size is not None: return chunk_size` and do the dask stuff one level of indentation less?

File: satpy/utils.py
Comment (djhoese) [inline satpy/utils.py:590]: Something about this logic being in two places in this module smells funny to me. If you think the code is cleaner with it this way I am OK with that. I just wanted to point it out.

File: satpy/utils.py
Comment (djhoese) [inline satpy/utils.py:577]: Are you using `float` versus `np.float32` or `np.float64` for a specific reason?

File: doc/source/faq.rst
Comment (djhoese) [inline doc/source/faq.rst:46]:
```suggestion
by using the following around your code:
```

File: doc/source/faq.rst
Comment (djhoese) [inline doc/source/faq.rst:57]: Would it be worth it to point out that a basic `dask.config.set(...)` at the top of their code would also accomplish this without having to use a context manager?

**Review by djhoese — CHANGES_REQUESTED** [review]
Wow, what a PR. Nice job getting all this sorted out. I have a couple questions and some suggestions that I've made inline. Otherwise, I have one major ask: For all the places you did `CHUNK_SIZE = get_legacy_chunk_size()` at the top level, how hard would it be to replace the in-function usage of `CHUNK_SIZE` with the `get_legacy_chunk_size()` function call? The goal being that users can use the context manager dask config setting to control their chunk sizes at runtime, not at import time.

**Comment by mraspaud** [comment]
> Otherwise, I have one major ask: For all the places you did `CHUNK_SIZE = get_legacy_chunk_size()` at the top level, how hard would it be to replace the in-function usage of `CHUNK_SIZE` with the `get_legacy_chunk_size()` function call? The goal being that users can use the context manager dask config setting to control their chunk sizes at runtime, not at import time.

I'm thinking this should be done at the reader level, ie that we should take the time and switch to data-native chunk size. The way we have it here, we're not changing anything in the readers at all. I was trying to preserve behaviour so to say, and was planning on inviting maintainers to go and revise each reader using the new chunk sizing concepts.

**Review by djhoese — APPROVED** [review]
I can be OK with that. Besides the pre-commit failure this looks good to me.

File: satpy/utils.py
Comment (mraspaud) [reply satpy/utils.py:626]: Re types: yes, we should have more info, but it seems like a longer text is needed, so probably we should just have it in the documentation. About that: we don't seem to have a dedicated section about chunking and its effects outside of the faq, don't you think it would be nice to have it somewhere?

Re the type of warning: I would love to have PendingDeprecationWarning, but I was thinking it would lack visibility?

File: satpy/utils.py
Comment (mraspaud) [reply satpy/utils.py]: Good idea. Done in [bf1c13d](https://github.com/pytroll/satpy/pull/2438/commits/bf1c13da26bda109b8e281e0d662cc693b9d3218)

File: satpy/utils.py
Comment (mraspaud) [reply satpy/utils.py:577]: To keep it generic, but I could change if want to.

File: satpy/utils.py
Comment (mraspaud) [reply satpy/utils.py]: Fixed, thanks

**Comment by gerritholl** [comment]
A couple of questions:

* Why?
* I currently have in `supervisord.conf` configured `PYTROLL_CHUNK_SIZE=4096`, which is then used when starting trollflow2 and thus when satpy is called.  The PR changes the documentation to say I should set this "in code" instead, but I don't have any custom code where I could put this.  How can I set chunk size in this case?

File: doc/source/faq.rst
Comment (mraspaud) [reply doc/source/faq.rst:57]: Good idea, but here again, I feel it should be its own separate section.

File: doc/source/faq.rst
Comment (gerritholl) [inline doc/source/faq.rst:66]: Can we please document a way to do this outside code?  I use this environment variable in my supervisor configuration, which calls trollflow2, which calls satpy.  This never reached custom code.

**Comment by mraspaud** [comment]
> A couple of questions:
> 
>     * Why?

Because I think having just one value like we have now is misleading. Having a size in bytes feels more correct. Also, I'm doing this to push for data-native chunk size, or rather chunk shapes, which are more compatible with providing the chunk size in bytes, rather than expecting square chunks with the way we have it now.

>     * I currently have in `supervisord.conf` configured `PYTROLL_CHUNK_SIZE=4096`, which is then used when starting trollflow2 and thus when satpy is called.  The PR changes the documentation to say I should set this "in code" instead, but I don't have any custom code where I could put this.  How can I set chunk size in this case?

I showed the python method because that is the regular way of setting config in dask, from what I understand. However, I believe I linked to the configuration documentation of dask, which shows how to use environment variables to achieve the same thing. In the current case, one can use the `DASK_ARRAY__CHUNK_SIZE` would be the variable to set.

File: doc/source/faq.rst
Comment (gerritholl) [inline doc/source/faq.rst:66]:
```suggestion
    The ``PYTROLL_CHUNK_SIZE`` variable is pending deprecation, so the
    above-mentioned dask configuration parameter should be used instead.
    You can also set the ``DASK_ARRAY__CHUNK_SIZE`` environment variable to
    the square of ``PYTROLL_CHUNK_SIZE``.
```

**Comment by gerritholl** [comment]
Could the PR description be a bit more elaborate as to why PYTROLL_CHUNK_SIZE is being deprecated?  I find the description of the PR a bit short considering the impact of the PR.

**Comment by mraspaud** [comment]
> Could the PR description be a bit more elaborate as to why PYTROLL_CHUNK_SIZE is being deprecated? I find the description of the PR a bit short considering the impact of the PR.

Yes, this became something else that what I had planned originally, hence the poor description. I'll update it, sorry for that.

File: satpy/utils.py
Comment (djhoese) [reply satpy/utils.py:628]: Documentation: That would be good in a restructure similar to what I'm doing in pyresample where dask arrays ("Chunked arrays") are their own section under a "Concepts" section of the documentation.

Warning: Yeah, true. Hard to see/find.

**Comment by djhoese** [comment]
Just to be sure, and I think I mentioned this in my inline comments in my last review, the dask options are all memory based, right? So a simple transition from `PYTROLL_CHUNK_SIZE` to the dask chunk size would likely take some consideration of the data type. Reader authors may need to be conscious/concerned with using the dask array chunk size as a guide and possibly tweaking it for different band resolutions. By that I mean, should a 500m and 1km band both have the same number of pixels per chunk? Or should we adjust the chunk size for one band to be a "resolution factor" compared to the other bands?

File: doc/source/faq.rst
Comment (mraspaud) [reply doc/source/faq.rst:59]: Ok, section added, please have a look.

File: doc/source/faq.rst
Comment (mraspaud) [reply doc/source/faq.rst]: Section added, please have a look.

**Comment by mraspaud** [comment]
> Just to be sure, and I think I mentioned this in my inline comments in my last review, the dask options are all memory based, right? So a simple transition from `PYTROLL_CHUNK_SIZE` to the dask chunk size would likely take some consideration of the data type. Reader authors may need to be conscious/concerned with using the dask array chunk size as a guide and possibly tweaking it for different band resolutions. By that I mean, should a 500m and 1km band both have the same number of pixels per chunk? Or should we adjust the chunk size for one band to be a "resolution factor" compared to the other bands?

Indeed, it will not be a 1 to 1 translation between the old and new parameters. My hope with this is that we can work with readers in a more flexible way. The resolution case you bring up is a good example of where we need to think about and test what is best for performance, but I think we should try to sick to what the user asked in the first place. Ie, one chunk at 1km resolution would cover 4 chunks at 500m resolution. We just need to make sure they are aligned as much as possible when we implement this in the readers.

**Comment by mraspaud** [comment]
> Could the PR description be a bit more elaborate as to why PYTROLL_CHUNK_SIZE is being deprecated? I find the description of the PR a bit short considering the impact of the PR.

Done

**Comment by mraspaud** [comment]
pre-commit.ci autofix

File: benchmarks/ahi_hsd_benchmarks.py
Comment (djhoese) [inline benchmarks/ahi_hsd_benchmarks.py:44]: Just curious, why do security checks flag this?

Comment (mraspaud) [reply benchmarks/ahi_hsd_benchmarks.py:44]: because `assert`

Comment (djhoese) [reply benchmarks/ahi_hsd_benchmarks.py:44]: Ah asserts outside of tests. Ok. Got it.

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:66]:
```suggestion
The usage of dask as the foundation for Satpy's operation means that the
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:67]: I'm not sure if there needs to be a semi-colon somewhere but a comma after this seems better...maybe?

```suggestion
underlying data is chunked, that is, cut in smaller pieces that can then be
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:69]:
```suggestion
processed in parallel. Information on dask's chunking can be found in the
dask documentation here: https://docs.dask.org/en/stable/array-chunks.html
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:73]:
```suggestion
Default chunk size used by Satpy can be configured by using the following
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:81]:
```suggestion
Or by using:
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:87]:
```suggestion
at the top of your code.
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:106]:
```suggestion
```

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:116]: This is not equivalent. This is only equivalent for 8-bit data types which is not the usual case for Satpy data.

Comment (mraspaud) [reply doc/source/overview.rst:114]: Indeed, fixing

File: doc/source/overview.rst
Comment (djhoese) [reply doc/source/overview.rst:114]: Looks good.

File: doc/source/overview.rst
Comment (djhoese) [inline doc/source/overview.rst:123]: Can you check the rendering of this in the HTML? I'm pretty sure you shouldn't have the double backticks because you have the `::` which makes the indented portion a literal block.

Comment (mraspaud) [reply doc/source/overview.rst:123]: Ah yes of course

**Review by djhoese — APPROVED** [review]
LGTM except for the linting failures

**Comment by mraspaud** [comment]
pre-commit.ci autofix

**Pull request merged**
